Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift

```diff
@@ -0,0 +1,124 @@
+            let key = try ASN1BitString(derEncoded: &nodes)
+
+            return SubjectPublicKeyInfo(algorithmIdentifier: algorithmIdentifier, key: key)
+        }
+    }
+
+    private init(algorithmIdentifier: RFC5480AlgorithmIdentifier, key: ASN1BitString) {
+        self.algorithmIdentifier = algorithmIdentifier
+        self.key = key
+    }
+
+    internal init(algorithmIdentifier: RFC5480AlgorithmIdentifier, key: [UInt8]) {
+        self.algorithmIdentifier = algorithmIdentifier
+        self.key = ASN1BitString(bytes: key[...])
+    }
+
+    func serialize(into coder: inout DER.Serializer, withIdentifier identifier: ASN1Identifier) throws {
+        try coder.appendConstructedNode(identifier: identifier) { coder in
+            try coder.serialize(self.algorithmIdentifier)
+            try coder.serialize(self.key)
+        }
+    }
+}
+
+struct RFC5480AlgorithmIdentifier: DERImplicitlyTaggable, Hashable {
+    static var defaultIdentifier: ASN1Identifier {
+        .sequence
+    }
+
+    var algorithm: ASN1ObjectIdentifier
+
+    var parameters: ASN1Any?
+
+    init(algorithm: ASN1ObjectIdentifier, parameters: ASN1Any?) {
+        self.algorithm = algorithm
+        self.parameters = parameters
+    }
+
+    init(derEncoded rootNode: ASN1Node, withIdentifier identifier: ASN1Identifier) throws {
+        // The AlgorithmIdentifier block looks like this.
+        //
+        // AlgorithmIdentifier  ::=  SEQUENCE  {
+        //   algorithm   OBJECT IDENTIFIER,
+        //   parameters  ANY DEFINED BY algorithm OPTIONAL
+        // }
+        //
+        // ECParameters ::= CHOICE {
+        //   namedCurve         OBJECT IDENTIFIER
+        //   -- implicitCurve   NULL
+        //   -- specifiedCurve  SpecifiedECDomain
+        // }
+        //
+        // We don't bother with helpers: we just try to decode it directly.
+        self = try DER.sequence(rootNode, identifier: identifier) { nodes in
+            let algorithmOID = try ASN1ObjectIdentifier(derEncoded: &nodes)
+
+            let parameters = nodes.next().map { ASN1Any(derEncoded: $0) }
+
+            return .init(algorithm: algorithmOID, parameters: parameters)
+        }
+    }
+
+    func serialize(into coder: inout DER.Serializer, withIdentifier identifier: ASN1Identifier) throws {
+        try coder.appendConstructedNode(identifier: identifier) { coder in
+            try coder.serialize(self.algorithm)
+            if let parameters = self.parameters {
+                try coder.serialize(parameters)
+            }
+        }
+    }
+}
+
+extension SubjectPublicKeyInfo {
+    static func stripRsaPssParameters(derEncoded: [UInt8]) throws -> [UInt8] {
+        guard var spki = try? SubjectPublicKeyInfo(derEncoded: derEncoded) else {
+            // If it's not a SPKI then it can't be a PSS key so we just return it.
+            return derEncoded
+        }
+
+        if spki.algorithmIdentifier.algorithm == .AlgorithmIdentifier.rsaPSS {
```

Comment:
Can we turn this into a shortcut early return? If we don't have to make any modifications, we don't need to re-encode the key.